File: 0078-subsets/0078-subsets.cpp
class Solution {
public:
    vector<vector<int>> subsets(vector<int>& nums) {
        vector<vector<int>> ans;
        vector<int> temp;
        recurse(nums, ans, temp, 0);
        return ans;
    }
    
    void recurse(vector<int>& nums, vector<vector<int>>& ans,vector<int> &temp, int index ) {
        if(index==nums.size()) {
            ans.push_back(temp);
            return;
        }
        // take
        temp.push_back(nums[index]);
        recurse(nums,ans,temp, index+1);
        
        //not take
        temp.pop_back();
        recurse(nums,ans,temp,index+1);
    }
};

File: 0078-subsets/0078-subsets.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var subsets = function(nums) {
    let ans = [], buffer = [];
    recurse(0);
    return ans;
    
    function recurse(index) {
        if(index==nums.length) {
            ans.push(buffer.slice());
            return;
        }
        buffer.push(nums[index]);
        recurse(index+1);
        buffer.pop();
        recurse(index+1);
    }
};